rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+*   Convergence tests: Run a curated set of small models until convergence. Measure the time to converge, and steps to converge
+*   Performance tests: Run a curated set of models for a pre-selected number of steps. Measure steps per second, and for image models images per second.
+
+
+## Releases
+
+In the multi package environment, we aim to have the following with the releases:
+
+1.  Smaller packages: We expect to release multiple small packages, instead of the giant package we are building and releasing right now.
+1.  Faster releases: We would like the smaller packages to be able to release much faster. With the ability to pin each dependency to a known good version, we would like to be able to compartmentalize issues, and only hold certain components back when only certain code in TF has issues.
+1.  Independent releases: With compatibility checks, we would like to be able to independently release different packages
+
+Below summarizes a timeline of releases for three packages, A, B and C, where B depends on A and C depends on B.
+
+
+![alt_text](20190305-modular-tensorflow/releases.png "Release plans")
+
+
+To summarize the above timeline:
+
+*   Different packages set their own release cadences
+*   Each package will set version boundaries for each of their dependencies.
+*   Each package is responsible for ensuring that all of their public APIs are working without any changes until the next major release
+*   Packages do not need to modify the minimum version requirements unless they start using newly introduced public API symbols.
+*   TF metapackage releases may choose to hold back individual packages in favor of faster releases. But dependency requirements have to be respected when doing so.
+*   Major releases still need to be coordinated.
+
+
+## Packaging Issue Scenarios
+
+
+We will now go over a few failure modes in the proposed environment, and propose how things need to be resolved. We will use a simple example, where we have three packages, A, B and C, where C depends on B and A, and B depends on A.
+
+
+#### Scenario 1: B does not work with a new release of A
+
+Potential issues:
+
+*   B uses non-public symbols from A:
+    *   B has to release a new version avoiding non-public APIs
+*   A changed public APIs:
+    *   A has to revert changes to its public APIs and release a patch version replacing the bad version
+
+
+#### Scenario 2: New release of B does not work with A
+
+Potential issues:
+
+*   B depends on unreleased APIs A exposed in a newer version
+    *   B needs to release a patch version, with just a change in the minimum required version of A
+
+
+#### Scenario 3: C and B depend on different minimum versions of A
+
+As both C and B have to define a range of versions they require from A, the max version should satisfy both constraints.
```

Comment:
How do plugins catch these dependency chain conflicts with sufficient time to resolve them before a release cycle?

Reply:
What kind of conflicts did you have in mind?
As described here, when we set the max required version to the next major release, plugins can release independently, and as each plugin is independently released pip should be able to take care of satisfying the constraints for 'A'. with the max required A version the same for all plugins, the requirement intervals of versions of 'A' are guaranteed to have an intersection.